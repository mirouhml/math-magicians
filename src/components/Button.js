import React from 'react';

export default class Button extends React.Component {
  constructor(props) {
    super(props);
    Object.assign(this, this.props);
  }

  render = () => {
    const row = this.buttons.map((cell, index) => {
      if (this.last) {
        if (index === 0) return <td key={cell} colSpan="2"><button type="button" onClick={() => this.action(cell)}>{cell}</button></td>;
        if (index === 2) return <td key={cell}><button className="operator" type="button" onClick={() => this.action(cell)}>{cell}</button></td>;
        return <td key={cell}><button type="button" onClick={() => this.action(cell)}>{cell}</button></td>;
      }
      if (index === 3) return <td key={cell}><button className="operator" type="button" onClick={() => this.action(cell)}>{cell}</button></td>;
      return <td key={cell}><button type="button" onClick={() => this.action(cell)}>{cell}</button></td>;
    });
    return <tr>{row}</tr>;
  };
}
