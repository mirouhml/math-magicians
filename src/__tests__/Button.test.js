import React from 'react';
import renderer from 'react-test-renderer';
import { screen, render, fireEvent } from '@testing-library/react';
import Button from '../components/Button';
import '@testing-library/jest-dom';

describe('Button component tests:', () => {
  const buttons = [
    ['AC', '+/-', '%', 'รท'],
    ['7', '8', '9', 'ร'],
    ['4', '5', '6', '-'],
    ['1', '2', '3', '+'],
    ['0', '.', '='],
  ];

  const onClick = jest.fn();

  it('Button component renders correctly', () => {
    const tree = renderer
      .create(<Button buttons={buttons[0]} action={onClick} last={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Button click interaction - test onclick \'+/-\'', () => {
    render(<Button buttons={buttons[0]} action={onClick} last={false} />);
    fireEvent.click(screen.getByText('+/-'));
    fireEvent.click(screen.getByText('+/-'));
    expect(onClick).toHaveBeenCalledTimes(2);
  });

  it('Button click interaction - test onclick \'9\'', () => {
    render(<Button buttons={buttons[1]} action={onClick} last={false} />);
    fireEvent.click(screen.getByText('9'));
    fireEvent.click(screen.getByText('9'));
    fireEvent.click(screen.getByText('9'));
    expect(onClick).toHaveBeenCalledTimes(3);
  });

  it('Button click interaction - test onclick \'-\'', () => {
    render(<Button buttons={buttons[2]} action={onClick} last={false} />);
    fireEvent.click(screen.getByText('-'));
    fireEvent.click(screen.getByText('-'));
    expect(onClick).toHaveBeenCalledTimes(2);
  });

  it('Button click interaction - test onclick \'2\'', () => {
    render(<Button buttons={buttons[3]} action={onClick} last={false} />);
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('2'));
    expect(onClick).toHaveBeenCalledTimes(4);
  });

  it('Button click interaction - test onclick \'=\'', () => {
    render(<Button buttons={buttons[4]} action={onClick} last />);
    fireEvent.click(screen.getByText('='));
    fireEvent.click(screen.getByText('='));
    expect(onClick).toHaveBeenCalledTimes(2);
  });
});
