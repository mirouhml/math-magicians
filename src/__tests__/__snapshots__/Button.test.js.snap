// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests: Button component renders correctly 1`] = `
<tr>
  <td>
    <button
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
  </td>
  <td>
    <button
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
  </td>
  <td>
    <button
      onClick={[Function]}
      type="button"
    >
      %
    </button>
  </td>
  <td>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
  </td>
</tr>
`;
